lib/membrane_rtsp_plugin/source.ex

```diff
@@ -90,93 +109,93 @@ defmodule Membrane.RTSP.Source do
+  def handle_child_notification(:ready, :ready_notifier, _ctx, state) do
+    send(state.connection_manager, :source_ready)
     {[], state}
   end
 
   @impl true
-  def handle_pad_added(Pad.ref(:output, ssrc) = pad, _ctx, state) do
-    track = Map.fetch!(state.ssrc_to_track, ssrc)
-
-    spec =
-      get_child(:rtp_session)
-      |> via_out(Pad.ref(:output, ssrc), options: [depayloader: get_rtp_depayloader(track)])
-      |> parser(track)
-      |> bin_output(pad)
-
-    {[spec: spec], state}
+  def handle_child_notification(_notification, _element, _ctx, state) do
+    {[], state}
   end
 
   @impl true
-  def handle_info({:tracks, tracks}, _ctx, state) do
-    Membrane.Logger.info("Received tracks: #{inspect(tracks)}")
-
+  def handle_info(%{tracks: tracks, transport_info: transport_info}, _ctx, state) do
     fmt_mapping =
       Enum.map(tracks, fn %{rtpmap: rtpmap} ->
         {rtpmap.payload_type, {String.to_atom(rtpmap.encoding), rtpmap.clock_rate}}
       end)
       |> Enum.into(%{})
 
-    case state.transport do
-      :tcp ->
-        local_socket = List.first(tracks).transport
+    common_spec = [
+      child(:rtp_session, %Membrane.RTP.SessionBin{fmt_mapping: fmt_mapping}),
+      child(:ready_notifier, ReadyNotifier)
+    ]
 
+    case transport_info do
+      {:tcp, tcp_socket} ->
         spec =
-          child(:source, %TCP{local_socket: local_socket})
-          |> child(:tcp_depayloader, Decapsulator)
-          |> via_in(Pad.ref(:rtp_input, make_ref()))
-          |> child(:rtp_session, %Membrane.RTP.SessionBin{fmt_mapping: fmt_mapping})
+          common_spec ++
+            [
+              child(:tcp_source, %Membrane.TCP.Source{
+                connection_side: :client,
+                local_socket: tcp_socket
+              })
+              |> child(:tcp_depayloader, Decapsulator)
+              |> via_in(Pad.ref(:rtp_input, make_ref()))
+              |> get_child(:rtp_session)
+            ]
 
-        {[spec: spec], %{state | tracks: tracks, tcp_socket: local_socket}}
+        {[spec: spec], %{state | tracks: tracks}}
 
-      :udp ->
+      {:udp, min_port, max_port} ->
         spec =
-          [child(:rtp_session, %Membrane.RTP.SessionBin{fmt_mapping: fmt_mapping})] ++
-            Enum.flat_map(tracks, fn track ->
-              {rtp_port, rtcp_port} = track.transport
-
-              [
-                child({:udp, make_ref()}, %Membrane.UDP.Source{local_port_no: rtp_port})
-                |> via_in(Pad.ref(:rtp_input, make_ref()))
-                |> get_child(:rtp_session),
-                child({:udp, make_ref()}, %Membrane.UDP.Source{local_port_no: rtcp_port})
-                |> via_in(Pad.ref(:rtp_input, make_ref()))
-                |> get_child(:rtp_session)
-              ]
+          common_spec ++
+            Enum.map(min_port..max_port, fn port ->
+              child({:udp_source, make_ref()}, %Membrane.UDP.Source{local_port_no: port})
```

Comment:
Why do we listen on all the ports from the range?

Reply:
this range includes only ports negotiated with setup messages, see connection_manager.ex:147, I'll make it more readable

Reply:
bump @Noarkhh

Reply:
There was this slight concern, that the original version checked for a free port only during the SETUP phase and didn't claim it, so theoretically the port could be taken by some other process after SETUP and before the UDP sources are spawned. To eliminate this possibility entirely we would need to add an option to UDP Source to pass an already connected socket to it, but that could complicate the code a bit.